File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/DayOneViewModel.java
package droiddevelopers254.droidconke.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;

import droiddevelopers254.droidconke.datastates.SessionsState;
import droiddevelopers254.droidconke.repository.DayOneRepo;
import droiddevelopers254.droidconke.repository.RoomStarrSessionRepo;

public class DayOneViewModel extends ViewModel {
    private MediatorLiveData<SessionsState> sessionsStateMediatorLiveData;
    private DayOneRepo dayOneRepo;
    private RoomStarrSessionRepo roomStarrSessionRepo;

    public DayOneViewModel(){
        dayOneRepo= new DayOneRepo();
        sessionsStateMediatorLiveData = new MediatorLiveData<>();
        roomStarrSessionRepo = new RoomStarrSessionRepo();
    }

    public LiveData<SessionsState> getSessions(){
        return sessionsStateMediatorLiveData;
    }

    public void getDayOneSessions(){
        final LiveData<SessionsState> sessionsStateLiveData = dayOneRepo.getDayOneSessions();
        sessionsStateMediatorLiveData.addSource(sessionsStateLiveData,
                sessionsStateMediatorLiveData ->{
            if (this.sessionsStateMediatorLiveData.hasActiveObservers()){
                this.sessionsStateMediatorLiveData.removeSource(sessionsStateLiveData);
            }
            this.sessionsStateMediatorLiveData.setValue(sessionsStateMediatorLiveData);
                });
    }
}

File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/DayOneViewModel.kt
package droiddevelopers254.droidconke.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import droiddevelopers254.droidconke.datastates.Result
import droiddevelopers254.droidconke.models.SessionsModel
import droiddevelopers254.droidconke.repository.DayOneRepo
import droiddevelopers254.droidconke.utils.NonNullMediatorLiveData
import kotlinx.coroutines.launch

class DayOneViewModel(private val dayOneRepo: DayOneRepo) : ViewModel() {
  private val sessionsStateMediatorLiveData = NonNullMediatorLiveData<List<SessionsModel>>()
  private val sessionError = NonNullMediatorLiveData<String>()


  fun getSessionsResponse(): LiveData<List<SessionsModel>> = sessionsStateMediatorLiveData

  fun getSessionsError(): LiveData<String> = sessionError

  fun getDayOneSessions() {
    viewModelScope.launch {
      when (val value = dayOneRepo.getDayOneSessions()) {
          is Result.Success -> sessionsStateMediatorLiveData.postValue(value.data)
          is Result.Error -> sessionError.postValue(value.exception)
      }
    }
  }
}
